mailbox/store/src/main/java/org/apache/james/mailbox/store/StoreRightManager.java

```diff
@@ -161,12 +163,45 @@ public void setRights(MailboxId mailboxId, MailboxACL mailboxACL, MailboxSession
 
     @Override
     public void setRights(MailboxPath mailboxPath, MailboxACL mailboxACL, MailboxSession session) throws MailboxException {
+        validateDomainsAreIdentical(mailboxPath.getUser(), mailboxACL);
         MailboxMapper mapper = mailboxSessionMapperFactory.getMailboxMapper(session);
         Mailbox mailbox = mapper.findMailboxByPath(mailboxPath);
 
         setRights(mailboxACL, mapper, mailbox);
     }
 
+    @VisibleForTesting
+    void validateDomainsAreIdentical(String user, MailboxACL mailboxACL) throws DifferentDomainException {
+        if (mailboxACL.getEntries().keySet().stream()
+            .filter(entry -> !entry.getNameType().equals(NameType.special))
+            .map(EntryKey::getName)
+            .anyMatch(name -> areDomainsDifferent(name, user))) {
+            throw new DifferentDomainException();
+        }
+    }
+
+    @VisibleForTesting
+    boolean areDomainsDifferent(String user, String otherUser) {
+        Optional<String> domain = extractDomain(user);
+        Optional<String> otherDomain = extractDomain(otherUser);
+        if (domain.isPresent() != otherDomain.isPresent()) {
+            return true;
+        }
+        if (!domain.isPresent()) {
+            return false;
+        }
+        return !domain.get().equals(otherDomain.get());
+    }
+
+    @VisibleForTesting 
+    Optional<String> extractDomain(String user) {
+        String[] split = user.split("@");
```

Comment:
the 12432th implementation of domain extraction

Reply:
Yes !!!